packages/flutter/lib/src/material/autocomplete.dart

```diff
@@ -0,0 +1,256 @@
+
+  /// {@macro flutter.widgets.RawAutocomplete.onSelected}
+  final AutocompleteOnSelected<T>? onSelected;
+
+  /// {@macro flutter.widgets.RawAutocomplete.optionsBuilder}
+  final AutocompleteOptionsBuilder<T> optionsBuilder;
+
+  /// {@macro flutter.widgets.RawAutocomplete.optionsViewBuilder}
+  ///
+  /// If not provided, will build a standard Material-style list of results by
+  /// default.
+  final AutocompleteOptionsViewBuilder<T>? optionsViewBuilder;
+
+  static Widget _defaultFieldViewBuilder(BuildContext context, TextEditingController textEditingController, FocusNode focusNode, VoidCallback onFieldSubmitted) {
+    return _AutocompleteField(
+      focusNode: focusNode,
+      textEditingController: textEditingController,
+      onFieldSubmitted: onFieldSubmitted,
+    );
+  }
+
+  @override
+  _AutocompleteState<T> createState() => _AutocompleteState<T>();
+}
+
+class _AutocompleteState<T extends Object> extends State<Autocomplete<T>> {
+  @override
+  Widget build(BuildContext context) {
+    return RawAutocomplete<T>(
+      displayStringForOption: widget.displayStringForOption,
+      fieldViewBuilder: widget.fieldViewBuilder,
+      optionsBuilder: widget.optionsBuilder,
+      optionsViewBuilder: widget.optionsViewBuilder ?? (BuildContext context, AutocompleteOnSelected<T> onSelected, Iterable<T> options) {
+        return _AutocompleteOptions<T>(
+          displayStringForOption: widget.displayStringForOption,
+          onSelected: onSelected,
+          options: options,
+        );
+      },
+      onSelected: widget.onSelected,
+    );
+  }
+}
+
+// The default Material-style Autocomplete text field.
+class _AutocompleteField extends StatelessWidget {
+  const _AutocompleteField({
+    Key? key,
+    required this.focusNode,
+    required this.textEditingController,
+    required this.onFieldSubmitted,
+  }) : super(key: key);
+
+  final FocusNode focusNode;
+
+  final VoidCallback onFieldSubmitted;
+
+  final TextEditingController textEditingController;
+
+  @override
+  Widget build(BuildContext context) {
+    return TextFormField(
+      controller: textEditingController,
+      focusNode: focusNode,
+      onFieldSubmitted: (String value) {
+        onFieldSubmitted();
+      },
+    );
+  }
+}
+
+// The default Material-style Autocomplete options.
+class _AutocompleteOptions<T extends Object> extends StatelessWidget {
+  const _AutocompleteOptions({
+    Key? key,
+    required this.displayStringForOption,
+    required this.onSelected,
+    required this.options,
+  }) : super(key: key);
+
+  final AutocompleteOptionToString<T> displayStringForOption;
+
+  final AutocompleteOnSelected<T> onSelected;
+
+  final Iterable<T> options;
+
+  @override
+  Widget build(BuildContext context) {
+    return Align(
+      alignment: Alignment.topLeft,
+      child: Material(
+        elevation: 4.0,
+        child: Container(
+          height: 200.0,
+          child: ListView.builder(
```

Comment:
ListView behaves funky when the TextField itself is constrained:

<img width="722" alt="Screen Shot 2021-01-21 at 3 58 46 PM" src="https://user-images.githubusercontent.com/27032613/105320794-90dc2000-5c01-11eb-8d8e-f88d41442550.png">

<details><summary>Sample app</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(
              width: 200,
              child: AutocompleteBasicExample(),
            ),
          ],
        ),
      ),
    );
  }
}

class AutocompleteBasicExample extends StatelessWidget {
  AutocompleteBasicExample({Key key}) : super(key: key);

  final List<String> _kOptions = <String>[
    'aardvark',
    'bobcat',
    'chameleon',
    'dingo',
    'elephant',
    'flamingo',
    'goose',
    'hippopotamus',
    'iguana',
    'jaguar',
    'koala',
    'lemur',
    'mouse',
    'northern white rhinocerous',
  ];

  @override
  Widget build(BuildContext context) {
    return Autocomplete<String>(
      optionsBuilder: (TextEditingValue textEditingValue) {
        if (textEditingValue.text == '') {
          return const Iterable<String>.empty();
        }
        return _kOptions.where((String option) {
          return option.contains(textEditingValue.text.toLowerCase());
        });
      },
      optionsViewBuilder: (
        BuildContext context,
        AutocompleteOnSelected<String> onSelected,
        Iterable<String> options,
      ) {
        return Align(
          alignment: Alignment.topLeft,
          child: Material(
            elevation: 4.0,
            child: Container(
              height: 200.0,
              child: ListView.builder(
                padding: const EdgeInsets.all(8.0),
                itemCount: options.length,
                itemBuilder: (BuildContext context, int index) {
                  final String option = options.elementAt(index);
                  return InkWell(
                    onHover: (bool enableFeedback) {},
                    onTap: () {
                      onSelected(option);
                    },
                    child: Container(
                      constraints: const BoxConstraints(minHeight: 24.0),
                      alignment: AlignmentDirectional.centerStart,
                      child: Text(option),
                    ),
                  );
                },
              ),
            ),
          ),
        );
      },
      onSelected: (String selection) {
        print('You just selected $selection');
      },
    );
  }
}
```

</details>

Basically, the ListView itself extends beyond the text field's length when it should end. I wonder if there's an easy way to figure the width of the text field out and limit it accordingly?

Reply:
I still haven't taken a look at this but I'll try to come back to it later today.  I thought I made an example of this at some point, but maybe I broke it again with all the changes or something.

Reply:
The example I was thinking of ([here](https://github.com/justinmc/flutter-autocomplete-examples/blob/master/lib/core_basic_narrow_page.dart)), just manually sets the width of the optionsViewBuilder to the same as the field.  Hans was pushing for the ability to handle this case, and this is what I came up with for the short term.

But I really feel like I might be able to get the width of the field using the FocusNode and make this work nicely right now...  I'll take another closer look tomorrow and see, otherwise I'll postpone it for a later PR.

Reply:
I played around with a bunch of different options for this, and my conclusion is that we should come back to this in another PR if you're ok with that.  I've created an issue to discuss our options: https://github.com/flutter/flutter/issues/74509

The good news is that it should be possible to do anything we want here.  I was able to get the size of the field via the FocusNode (even when the field is not created with fieldViewBuilder).  I was also able to get the constraints of the Autocomplete widget at its point in the tree using a LayoutBuilder.

The problem is I'm not fully satisfied with any one solution here.  I could set the width to the width of the field, but some users may want it to be full-width or the width of the Autocomplete instead (e.g. split field and options).  This is quickly infringing on the idea of flat options, which I mentioned in the design doc and we should also support at some point.

So, I think we should leave the behavior the same as RawAutocomplete for now and come back to this with a full solution.  I've proposed another builder method that would allow full customizability (see the issue linked above).

Let me know what you think.  For now, if users want to achieve a narrow field with results that match, they would need to manually set the width of each to be the same, as it is with RawAutocomplete (that's how it's done in the [example I was thinking of](https://github.com/justinmc/flutter-autocomplete-examples/blob/master/lib/core_basic_narrow_page.dart)).

Reply:
That makes sense! Happy to move forward with what we have for now and come back to this in a later PR.